src/lease.ts

```diff
@@ -0,0 +1,255 @@
+import { EventEmitter } from 'events';
+
+import { ConnectionPool } from './connection-pool';
+import { castGrpcError, EtcdError, EtcdLeaseInvalidError, GRPCConnectFailedError } from './errors';
+import { PutBuilder } from './kv-builder';
+import * as RPC from './rpc';
+
+function throwIfError<T>(value: T | Error): T {
+  if (value instanceof Error) {
+    throw value;
+  }
+
+  return value;
+}
+
+function leaseExpired(lease: RPC.ILeaseKeepAliveResponse) {
+  return lease.TTL === '0';
+}
+
+/**
+ * Implements RPC.ICallable. Wraps a pool and adds the `leaseID` to outgoing
+ * put requests before executing them.
+ */
+class LeaseClientWrapper implements RPC.ICallable {
+  constructor(
+    private leaseID: Promise<string | Error>,
+    private pool: ConnectionPool,
+  ) {}
+
+  public exec(service: keyof typeof RPC.Services, method: string, payload: any): Promise<any> {
```

Comment:
DT recommends against that style in their "common mistakes": 

![](https://peet.io/i/17-04-g2w2fid9w96iuo0hoxo5u9wghgizvz.png)